File: Python/UniqueBinarySearchTrees_II.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def generateTrees(self, n: int) -> list[Optional[TreeNode]]:
        if n == 0:
            return []

        memo = dict()

        def build(start, end):

            key = f"{start}={end}"

            if key in memo:
                return memo.get(key)

            trees = []

            if start > end:
                trees.append(None)
                return trees

            for root in range(start, end + 1):
                leftTrees = build(start, root - 1)
                rightTrees = build(root + 1, end)

                for left in leftTrees:
                    for right in rightTrees:
                        node = TreeNode(root)
                        node.left = left
                        node.right = right
                        trees.append(node)

            memo.setdefault(key, trees)
            return trees

        return build(1, n)

File: JavaScript/UniqueBinarySearchTrees_II.js
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

const generateTrees = function (n) {
  if (n === 0) return [];

  const memo = new Map();

  function build(start, end) {
    const key = `${start}-${end}`;

    if (memo.has(key)) return memo.get(key);

    const tree = [];

    if (start > end) {
      tree.push(null);
      return tree;
    }

    for (let root = start; root <= end; root++) {
      const leftTrees = build(start, root - 1);
      const rightTrees = build(root + 1, end);

      for (let left of leftTrees) {
        for (let right of rightTrees) {
          const node = new TreeNode(root);
          node.left = left;
          node.right = right;
          tree.push(node);
        }
      }
    }
    memo.set(key, tree);
    return tree;
  }
  return build(1, n);
};
